packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts

```diff
@@ -458,4 +458,166 @@ describe('portfolio associations and product constraints', () => {
+    });
+    const executionRole = new iam.Role(stack, 'ExecutionRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+
+    portfolio.addProduct(product);
+
+    portfolio.deployWithStackSets(product, accounts, regions, adminRole, executionRole, true, {
+      description: 'stackset description',
+      messageLanguage: servicecatalog.MessageLanguage.JP,
+    });
+
+    expect(stack).toHaveResourceLike('AWS::ServiceCatalog::StackSetConstraint', {
+      PortfolioId: { Ref: 'MyPortfolio59CCA9C9' },
+      ProductId: { Ref: 'MyProduct49A3C587' },
+      AdminRole: {
+        'Fn::GetAtt': [
+          'AdminRole38563C57',
+          'Arn',
+        ],
+      },
+      ExecutionRole: { Ref: 'ExecutionRole605A040B' },
+      Description: 'stackset description',
+      AccountList: accounts,
+      RegionList: regions,
+      StackInstanceControl: 'ALLOWED',
+      AcceptLanguage: 'jp',
+    });
+  }),
+
+  test('deployment with stacksets still adds without explicit association', () => {
+    const accounts = ['012345678901', '012345678901'];
+    const regions = ['us-east-1', 'us-west-2', 'eu-west-1'];
+
+    const adminRole = new iam.Role(stack, 'AdminRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+    const executionRole = new iam.Role(stack, 'ExecutionRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+
+    portfolio.deployWithStackSets(product, accounts, regions, adminRole, executionRole, false);
+
+    expect(stack).toHaveResourceLike('AWS::ServiceCatalog::StackSetConstraint');
+  }),
+
+  test('fails to add multiple deploy with stackset constraints', () => {
+    const accounts = ['012345678901', '012345678901'];
+    const regions = ['us-east-1', 'us-west-2', 'eu-west-1'];
+
+    const adminRole = new iam.Role(stack, 'AdminRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+    const executionRole = new iam.Role(stack, 'ExecutionRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+
+    portfolio.deployWithStackSets(product, accounts, regions, adminRole, executionRole, true);
+
+    expect(() => {
+      portfolio.deployWithStackSets(product, accounts, regions, adminRole, executionRole, true);
+    }).toThrowError(/Cannot configure multiple stackset deployment constraints for association/);
+  }),
+
+  test('fails to configure deployment with stacksets if a launch role has been set', () => {
+    const launchRole = new iam.Role(stack, 'LaunchRole', {
+      assumedBy: new iam.ServicePrincipal('servicecatalog.amazonaws.com'),
+    });
+
+    const accounts = ['012345678901', '012345678901'];
+    const regions = ['us-east-1', 'us-west-2', 'eu-west-1'];
+
+    const adminRole = new iam.Role(stack, 'AdminRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+    const executionRole = new iam.Role(stack, 'ExecutionRole', {
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+
+    portfolio.setLaunchRole(product, launchRole);
```

Comment:
Feels like there's some duplication in these tests.

How about we have a nested `describes` that has a `beforeEach()` like this:

```ts
describes('a Product with a launch Role constraint', () => {

  beforeEach(() => {
    const launchRole = new iam.Role(stack, 'LaunchRole', {
      assumedBy: new iam.ServicePrincipal('servicecatalog.amazonaws.com'),
    });
    portfolio.setLaunchRole(product, launchRole);
  });

  test('creates the correct resource', () => {
     // positive tests case...
  });

  test('does not allow adding a StackSet constraint to the same Portfolio-Product pair', () => {
    const adminRole = new iam.Role(stack, 'AdminRole', {
      assumedBy: new iam.AccountRootPrincipal(),
    });
    const executionRole = new iam.Role(stack, 'ExecutionRole', {
      assumedBy: new iam.AccountRootPrincipal(),
    });

     expect(() => {
      portfolio.deployWithStackSets(product, accounts, regions, adminRole, executionRole, false);
    }).toThrowError(/Cannot configure stackset deployment when a launch role is already defined for association/);
  });
});
```

Something like this. Should make these cleaner.

Reply:
What would be nice would be if we could define role(s) in the beforeEach, but the linting rules make it such that cannot have any unused ones.  Is it possible to disable that for this?  Otherwise I'm not sure how convenient this is for just one/two tests.  If I could define a `launchRole`, `adminRole`, and `executionRole` (but not associate them) then it would make all of these substantially cleaner.  Something like
```
  beforeEach(() => {
    const launchRole = new iam.Role(stack, 'LaunchRole', {
      assumedBy: new iam.ServicePrincipal('servicecatalog.amazonaws.com'),
    });
    const adminRole = new iam.Role(stack, 'AdminRole', {
      assumedBy: new iam.AccountRootPrincipal(),
    });
    const executionRole = new iam.Role(stack, 'ExecutionRole', {
      assumedBy: new iam.AccountRootPrincipal(),
    });
    // Add some lines to get around declared variables being unused 
  });
```
Or would that be antithetical to jest testing.

Reply:
This doesn't work?

```ts
  let launchRole: iam.IRole, adminRole: iam.IRole, executionRole: iam.IRole;
  beforeEach(() => {
    launchRole = new iam.Role(stack, 'LaunchRole', {
      assumedBy: new iam.ServicePrincipal('servicecatalog.amazonaws.com'),
    });
    adminRole = new iam.Role(stack, 'AdminRole', {
      assumedBy: new iam.AccountRootPrincipal(),
    });
    executionRole = new iam.Role(stack, 'ExecutionRole', {
      assumedBy: new iam.AccountRootPrincipal(),
    });
    // Add some lines to get around declared variables being unused 
  });
```

Reply:
I embarrassingly left `const` in which was actual root of issue.  This is great now really DRYs it up.